kb: (cve-2025-1974) document steps to preserve default and any custom configuration (#81)
harvester/harvesterhci.io: kb/2025-03-25/cve-2025-1974.md

If the command returns one of the affected versions, disable the `rke2-ingress-nginx-admission` validating webhook configuration by performing the following steps:

1. On one of your control plane nodes, use `kubectl` to confirm the existence of the `HelmChartConfig` resource named `rke2-ingress-nginx`:

   ```sh
   $ kubectl -n kube-system get helmchartconfig rke2-ingress-nginx
   NAME                 AGE
   rke2-ingress-nginx   14d1h
   ```

> 1. Use `kubectl -n kube-system edit helmchartconfig rke2-ingress-nginx` to add the following configurations to the resource:

Author: `kubectl patch` won't work here because the value of `.spec.valuesContent` is a multi-line string blob.

          enabled: false
        extraArgs:
          enable-annotation-validation: true

>           default-ssl-certificate: cattle-system/tls-rancher-internal

Reviewer: In Harvester v1.5.0-rc2, there is below config after system installation. 

```
$kubectl get helmchartconfig -A
NAMESPACE     NAME                 AGE
kube-system   rke2-ingress-nginx   2d

$kubectl get helmchartconfig -A -oyaml
apiVersion: v1
items:
- apiVersion: helm.cattle.io/v1
  kind: HelmChartConfig
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration:...
    creationTimestamp: "2025-03-25T18:23:35Z"
    generation: 2
    name: rke2-ingress-nginx
    namespace: kube-system
    resourceVersion: "6387"
    uid: 694c62b9-d27f-4243-bb1b-1747f2b027fd
  spec:
    valuesContent: |
      controller:
          admissionWebhooks:
              port: 8444
          config:
              proxy-body-size: "0"
              proxy-request-buffering: "off"
          extraArgs:
              default-ssl-certificate: cattle-system/tls-rancher-internal
          publishService:
              pathOverride: kube-system/ingress-expose
...
```

could you add more details about:
(1) Will this config yaml file be merged into the existing `rke2-ingress-nginx` ?  or replace current; 

(2) How to validate that: `Harvester automatically applies the change once the content is saved` ?

suggest user can to the output of `kubectl get helmchartconfig -A` first, add a new yaml file then, check `kubectl get helmchartconfig -A` again to ensure that the values are merged/replaced.

(3) If a user do `kubectl apply -f` instead of putting a yaml file to the path, will it work? or tell user explicitly that this is not working.

Thanks.

Author: It will replace the current configuration. From what I can tell, the default configuration is encrypted within the OOB `/var/lib/rancher/rke2/server/manifests/rke2-ingress-nginx.yaml` file. Instead of asking users to try to decrypt and monkey with that, we follow RKE2 KB to add a "drop-in" patch. Let me know if you are aware of a way to merge only the CVE patch.

As for the 3rd point, I lean towards including only what user needs to do, and avoid describing other options, especially those that don't work. I don't want to go into, "what about helm, kustomize, kubectl [create|patch|apply]".

Reviewer: FYI:

I did a test: set the `spec. valuesContent.controller.admissionWebhooks.enabled=false` manually, after reboot the cluster, the `helmchartconfig` object did not change, the change persistant.

```
$ kubectl get helmchartconfig -A -oyaml

apiVersion: v1
items:
- apiVersion: helm.cattle.io/v1
  kind: HelmChartConfig
  metadata:
    annotations:
      kubectl.kubernetes.io/last-applied-configuration: |
        {"apiVersion":"helm.cattle.io/v1","kind":"HelmChartConfig","metadata":{"annotations":{},"name":"rke2-ingress-nginx","namespace":"kube-system"},"spec":{"valuesContent":"controller:\n  config:\n    proxy-body-size: \"0\"\n    proxy-request-buffering: \"off\"\n  admissionWebhooks:\n    port: 8444\n  publishService:\n    pathOverride: kube-system/ingress-expose\n  extraArgs:\n    default-ssl-certificate: cattle-system/tls-rancher-internal"}}
    creationTimestamp: "2025-03-25T18:23:35Z"
    generation: 3
    name: rke2-ingress-nginx
    namespace: kube-system
    resourceVersion: "481126"
    uid: 694c62b9-d27f-4243-bb1b-1747f2b027fd
  spec:
    valuesContent: |
      controller:
          admissionWebhooks:
              port: 8444
              enabled: false   /// set via kubectl
          config:
              proxy-body-size: "0"
              proxy-request-buffering: "off"
          extraArgs:
              default-ssl-certificate: cattle-system/tls-rancher-internal
          publishService:
              pathOverride: kube-system/ingress-expose
```

Reviewer: @brandond I have no idea yet, how `HelmChartConfig` comes, but on Harvester cluster, the newly installed cluster has it :)

Reviewer: @ihcsim 

The file based solution, has another potential drawback: node is down/removed, new node is prompted and has no such file ...

Suppose Harvester is luckily to have this `HelmChartConfig` from the beginning, then Harvester's patch solution can be based on `kubectl`, it runs everywhere and persists.

Author: The Harvester's out-of-box `HelmChartConfig` comes from `/etc/rancher/rancherd/config.yaml.d/10-harvester.yaml` which don't get run during reboot. We keep only the `HelmChart` resources in `/var/lib/rancher/rke2/server/manifests/`. 

(@ibrokethecloud confirmed that `/etc/rancher/rancherd/config.yaml.d/10-harvester.yaml` does get re-run if the first node is re-initialized.)

@brandond If user added their own `HelmChartConfig` resources in `/var/lib/rancher/rke2/server/manifests/`, then any changes done via `kubectl` will get overridden, right? The only way around that is to merge the webhook changes into their existing `HelmChartConfig`.

@w13915984028 Agreed. The file-based solution isn't perfect. I was hoping to stay as close to the RKE2 CVE KB as possible. I am trying not to bloat the KB to account for all edges.

I pushed a commit with the `kubectl edit` approach and like it better. But if we can't guarantee its persistence, then we might have to use the file-base approach even though it's not perfect.

Author: I'm gonna stick with the approach documented in the RKE2 KB. Considering that this CVE affects a large portion of the Harvester user base, if any one of them later decided to add a `HelmChartConfig` resource into the rancher's manifests folder, it will override any changes made via `kubectl`. We can work with support on any edges that arise. Thanks!

          enabled: false
        extraArgs:
          enable-annotation-validation: true
          default-ssl-certificate: cattle-system/tls-rancher-internal
        config:
          proxy-body-size: "0"
          proxy-request-buffering: "off"
        publishService:
          pathOverride: kube-system/ingress-expose
  ```

    Harvester automatically applies the change once the content is saved.

> :::info important
> The above configuration will disable the RKE2 ingress-nginx admission webhooks, while preserving Harvester's default ingress-nginx configuration.

Reviewer 2: ```suggestion
    :::info important

    The configuration disables the RKE2 ingress-nginx admission webhooks while preserving Harvester's default ingress-nginx configuration.
```

> If other custom `rke2-ingress-nginx` configurations exist under the `/var/lib/rancher/rke2/server/manifests` folder, they must be merged with the above configuration under the same `HelmChartConfig` resource.
> :::

Reviewer 2: ```suggestion
    If other custom `rke2-ingress-nginx` configurations exist in the `/var/lib/rancher/rke2/server/manifests` folder, you must merge those with the above configuration under the same `HelmChartConfig` resource.

    :::
```

> Confirm that RKE2 deleted the `rke2-ingress-nginx-admission` validating webhook configuration:

Reviewer 2: ```suggestion
1. Verify that RKE2 deleted the `rke2-ingress-nginx-admission` validating webhook configuration.
```

```sh
$ kubectl get validatingwebhookconfiguration rke2-ingress-nginx-admission
Error from server (NotFound): validatingwebhookconfigurations.admissionregistration.k8s.io "rke2-ingress-nginx-admission" not found
```

> Ensure that the ingress-nginx pods are restarted successfully:

Reviewer 2: ```suggestion
1. Verify that the ingress-nginx pods are restarted successfully.
```

          enabled: false
        extraArgs:
          enable-annotation-validation: true
          default-ssl-certificate: cattle-system/tls-rancher-internal
        config:
          proxy-body-size: "0"
          proxy-request-buffering: "off"
        publishService:
          pathOverride: kube-system/ingress-expose
  ```

    Harvester automatically applies the change once the content is saved.

    :::info important

    The configuration disables the RKE2 ingress-nginx admission webhooks while preserving Harvester's default ingress-nginx configuration.

    If other custom `rke2-ingress-nginx` configurations exist in the `/var/lib/rancher/rke2/server/manifests` folder, you must merge those with the above configuration under the same `HelmChartConfig` resource.

>     :::

Reviewer 3: @jillian-maroket just noticed the indent is incorrect. Please help correct this. 

https://deploy-preview-81--harvester-home-preview.netlify.app/kb/2025/03/25/cve-2025-1974

Reviewer 2: @innobead The indentation should be correct because that note is part of the configuration step. I'm not sure why it's not displaying correctly.

Author: i added the fix to c9ff11818e4d323bf714f002000224097c2e2d91, along with an extra step to check for existing custom configuration.

          enabled: false
        extraArgs:
          enable-annotation-validation: true
          default-ssl-certificate: cattle-system/tls-rancher-internal
        config:
          proxy-body-size: "0"
          proxy-request-buffering: "off"
        publishService:
          pathOverride: kube-system/ingress-expose
  ```

    Harvester automatically applies the change once the content is saved.

:::info important

The configuration disables the RKE2 ingress-nginx admission webhooks while preserving Harvester's default ingress-nginx configuration.

If other custom `rke2-ingress-nginx` configurations exist, you must merge those with the above configuration under the same `HelmChartConfig` resource.

Use the `kubectl -n kube-system get helmchartconfig rke2-ingress-nginx -oyaml` command to check for existing custom configurations.

:::

1. Verify that RKE2 deleted the `rke2-ingress-nginx-admission` validating webhook configuration.

```sh
$ kubectl get validatingwebhookconfiguration rke2-ingress-nginx-admission
Error from server (NotFound): validatingwebhookconfigurations.admissionregistration.k8s.io "rke2-ingress-nginx-admission" not found
```

> 1. Verify that the ingress-nginx pods are restarted successfully.

Reviewer 4: The ordered list seems to be rendered incorrectly. Maybe the code block needs indentation?
<img width="672" alt="image" src="https://github.com/user-attachments/assets/d3e154ee-0e47-4908-a238-f2eeb56f955c" />

Reviewer 5: Seems need to indent the source code content
![image](https://github.com/user-attachments/assets/998222db-8270-4ddf-b5e6-9381b68761df)

    valuesContent: |-
      controller:
        admissionWebhooks:
          port: 8444
          enabled: false
        extraArgs:
          enable-annotation-validation: true
          default-ssl-certificate: cattle-system/tls-rancher-internal
        config:
          proxy-body-size: "0"
          proxy-request-buffering: "off"
        publishService:
          pathOverride: kube-system/ingress-expose
  ```

   Exit the `kubectl edit` command execution to save the configuration. 

   Harvester automatically applies the change once the content is saved.

>  :::info important

Author: This single-spaced indentation is needed to ensure the correct ordering of subsequent number bullets.
